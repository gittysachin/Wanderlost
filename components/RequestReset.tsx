import { useMutation, gql } from '@apollo/client'
import Form from './styles/Form'
import Error, { FormError } from './ErrorMessage'
import { useForm } from 'react-hook-form'
import {
  RequestReset as RequestResetTypes,
  RequestResetVariables,
} from './__generated__/RequestReset'

const REQUEST_RESET_MUTATION = gql`
  mutation RequestReset($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`

function RequestReset(): JSX.Element {
  const defaultValues = { email: '' }
  const { register, handleSubmit, reset: resetForm, errors } = useForm({
    defaultValues,
  })
  const [reset, { loading, error, called }] = useMutation<
    RequestResetTypes,
    RequestResetVariables
  >(REQUEST_RESET_MUTATION)

  async function onSubmit(data: RequestResetVariables) {
    await reset({
      variables: {
        email: data.email,
      },
    })
    resetForm(defaultValues)
  }

  return (
    <Form method="post" data-tested="form" onSubmit={handleSubmit(onSubmit)}>
      <fieldset disabled={loading} aria-busy={loading}>
        <h2>Request a password reset</h2>
        <Error error={error} />
        {!error && !loading && called && (
          <p>Success! Check your email for a reset link</p>
        )}
        <label>Email</label>
        <input type="email" name="email" ref={register({ required: true })} />
        <FormError error={errors.email} />

        <button type="submit">Request Reset!</button>
      </fieldset>
    </Form>
  )
}

export default RequestReset
export { REQUEST_RESET_MUTATION }
