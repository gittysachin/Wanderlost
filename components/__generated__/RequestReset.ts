/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: RequestReset
// ====================================================

export interface RequestReset_requestReset {
  __typename: 'SuccessMessage'
  message: string
}

export interface RequestReset {
  requestReset: RequestReset_requestReset
}

export interface RequestResetVariables {
  email: string
}
