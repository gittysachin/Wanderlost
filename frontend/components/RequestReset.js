import React, { useState } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { gql } from 'apollo-boost';
import Form from './styles/Form';
import Error from './ErrorMessage';

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

export default function Signin() {
  const [email, setEmail] = useState('');
  const [reset, { loading, error, called }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: {
        email
      }
    }
  );

  return (
    <Form
      method="post"
      onSubmit={async (e) => {
        e.preventDefault();
        await reset({ variables: { email } });
        setEmail('');
      }}
    >
      <fieldset disabled={loading} aria-busy={loading}>
        <h2>Request a password reset</h2>
        <Error error={error} />
        {!error && !loading && called && (
        <p>Success! Check your email for a reset link</p>
        )}
        <label htmlFor="email">
              Email
          <input
            type="email"
            name="email"
            placeholder="email"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
          />
        </label>

        <button type="submit">Request Reset!</button>
      </fieldset>
    </Form>
  );
}
