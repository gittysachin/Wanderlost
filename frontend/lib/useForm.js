import { useState } from 'react';

const uploadFile = async ({ value, inputs, updateInputs }) => {
  console.log('Uploading file');
  const data = new FormData();
  data.append('file', value);
  data.append('upload_preset', 'sickfits');

  const res = await fetch(
    'https://api.cloudinary.com/v1_1/davyhausser/image/upload',
    {
      method: 'POST',
      body: data,
    },
  );
  const file = await res.json();
  console.log(file);
  updateInputs({
    ...inputs,
    image: file.secure_url,
    largeImage: file.eager[0].secure_url,
  })
};

export default function useForm(initial = {}) {
  const [inputs, updateInputs] = useState(initial);

  function handleChange(e) {
    let { value, name, type } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      [value] = e.target.files;
      uploadFile({ value, inputs, updateInputs });
    }
    updateInputs({
      ...inputs,
      [name]: e.target.value,
    });
  }

  function resetForm() {
    updateInputs(initial);
  }

  return {
    inputs,
    handleChange,
    resetForm
  };
}