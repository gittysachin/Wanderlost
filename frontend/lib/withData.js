import withApollo from 'next-with-apollo';
import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
      request: (operation) => {
        operation.setContext({
          fetchOptions: {
            credentials: 'include',
          },
          headers,
        });
      },
    }),
    resolvers: {
      Mutation: {
        toggleCart(_, variables, { cache }) {
          // read the cartOpen value from the cache
          const { cartOpen } = cache.readQuery({
            query: LOCAL_STATE_QUERY,
          });
          // Write the cart state to the opposite
          const data = {
            data: { cartOpen: !cartOpen },
          };
          cache.writeData(data);
          return data;
        },
      },
    },
  });
}

export default withApollo(createClient);
