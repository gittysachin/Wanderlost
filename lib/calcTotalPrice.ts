import { GetCurrentUser_user_cart } from '../components/__generated__/GetCurrentUser'

export default function calcTotalPrice(
  cart: Array<GetCurrentUser_user_cart>
): number {
  return cart.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally
    return tally + cartItem.quantity * cartItem.item.price
  }, 0)
}
