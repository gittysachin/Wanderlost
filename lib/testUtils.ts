import '@testing-library/jest-dom/extend-expect'
import casual from 'casual'
import { GetItem_item } from '../components/__generated__/GetItem'
import { GetOrder_order, GetOrder_order_items } from '../components/__generated__/GetOrder'
import {
  GetCurrentUser_user,
  GetCurrentUser_user_cart,
} from '../components/__generated__/GetCurrentUser'

// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): GetItem_item => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
})

const fakeUser = (): Omit<GetCurrentUser_user, 'permissions'> => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  // permissions: [Permission.ADMIN],
  // orders: [],
  cart: [],
})

const fakeOrderItem = (): GetOrder_order_items => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = (): GetOrder_order => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
})

const fakeCartItem = (overrides: Partial<GetCurrentUser_user_cart>): GetCurrentUser_user_cart => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  // user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  private store: {
    [key: string]: string
  }

  constructor() {
    this.store = {}
  }

  clear(): void {
    this.store = {}
  }

  getItem(key: string): string | null {
    return this.store[key] || null
  }

  setItem(key: string, value: number): void {
    this.store[key] = value.toString()
  }

  removeItem(key: string): void {
    delete this.store[key]
  }
}

export { LocalStorageMock, fakeItem, fakeUser, fakeCartItem, fakeOrder, fakeOrderItem }
