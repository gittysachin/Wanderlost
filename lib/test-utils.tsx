import React from 'react'
import { render as defaultRender, RenderResult } from '@testing-library/react'
import { RouterContext } from 'next/dist/next-server/lib/router-context'
import { NextRouter } from 'next/router'
import '@testing-library/jest-dom/extend-expect'
import casual from 'casual'

export * from '@testing-library/react'

import { GetItem_item } from '../components/__generated__/GetItem'
import {
  GetOrder_order,
  GetOrder_order_items,
} from '../components/__generated__/GetOrder'
import {
  GetCurrentUser_user,
  GetCurrentUser_user_cart,
} from '../components/__generated__/GetCurrentUser'

// --------------------------------------------------
// Override the default test render with our own
//
// You can override the router mock like this:
//
// const { baseElement } = render(<MyComponent />, {
//   router: { pathname: '/my-custom-pathname' },
// });
// --------------------------------------------------
type DefaultParams = Parameters<typeof defaultRender>
type RenderUI = DefaultParams[0]
type RenderOptions = DefaultParams[1] & { router?: Partial<NextRouter> }

export function render(
  ui: RenderUI,
  { wrapper, router, ...options }: RenderOptions = {}
): RenderResult {
  if (!wrapper) {
    wrapper = ({ children }) => {
      const Wrapper = (
        <RouterContext.Provider value={{ ...mockRouter, ...router }}>
          {children}
        </RouterContext.Provider>
      )
      return Wrapper
    }
  }

  return defaultRender(ui, { wrapper, ...options })
}

const mockRouter: NextRouter = {
  basePath: '',
  pathname: '/',
  route: '/',
  asPath: '/',
  query: {},
  push: jest.fn(),
  replace: jest.fn(),
  reload: jest.fn(),
  back: jest.fn(),
  prefetch: jest.fn(),
  beforePopState: jest.fn(),
  events: {
    on: jest.fn(),
    off: jest.fn(),
    emit: jest.fn(),
  },
  isFallback: false,
}

// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): GetItem_item => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
})

const fakeUser = (): Omit<GetCurrentUser_user, 'permissions'> => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  // permissions: [Permission.ADMIN],
  // orders: [],
  cart: [],
})

const fakeOrderItem = (): GetOrder_order_items => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = (): GetOrder_order => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
})

const fakeCartItem = (
  overrides: Partial<GetCurrentUser_user_cart>
): GetCurrentUser_user_cart => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  // user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  private store: {
    [key: string]: string
  }

  constructor() {
    this.store = {}
  }

  clear(): void {
    this.store = {}
  }

  getItem(key: string): string | null {
    return this.store[key] || null
  }

  setItem(key: string, value: number): void {
    this.store[key] = value.toString()
  }

  removeItem(key: string): void {
    delete this.store[key]
  }
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
}
