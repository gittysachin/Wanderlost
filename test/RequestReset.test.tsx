import { render, userEvent } from '../lib/test-utils'
import RequestReset, {
  REQUEST_RESET_MUTATION,
} from '../components/RequestReset'

const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email: 'davy.hausser@gmail.com' },
    },
    result: {
      data: { requestReset: { message: 'success', __typename: 'Message' } },
    },
  },
]

describe('<RequestReset/>', () => {
  test('renders and matches snapshot', async () => {
    const { container } = render(<RequestReset />)
    expect(container).toMatchSnapshot()
  })

  test('calls the mutation', async () => {
    const {
      container,
      getByText,
      getByPlaceholderText,
      findByText,
    } = render(<RequestReset />, { mocks: mocks })

    userEvent.type(getByPlaceholderText('email'), 'davy.hausser@gmail.com')
    userEvent.click(getByText(/Request Reset/i))
    const success = await findByText(/Success/i)
    expect(success).toBeInTheDocument()
    const text = container.querySelector('p')?.innerHTML
    expect(text).toContain('Success! Check your email for a reset link!')
  })
})
