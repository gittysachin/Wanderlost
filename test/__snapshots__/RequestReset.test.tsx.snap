// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> renders and matches snapshot 1`] = `
<div>
  <form
    class="sc-bdnylx eviPWj"
    data-tested="form"
    method="post"
  >
    <fieldset
      aria-busy="false"
    >
      <h2>
        Request a password reset
      </h2>
      <label>
        Email
        <input
          name="email"
          placeholder="email"
          type="email"
        />
      </label>
      <button
        type="submit"
      >
        Request Reset!
      </button>
    </fieldset>
  </form>
</div>
`;
